File: SCimage/src/data/outputSize.ts
import {
  AUTO_OPTION,
  DEFAULT_OUTPUT_PROFILE_ID,
  OUTPUT_PROFILE_ASPECT_V1,
  OUTPUT_PROFILE_PIXEL_V1,
  aspectLabels,
  parsePixelSize,
  pixelSizeValues,
  type OutputOption,
} from "./outputProfiles";
import { getResolvedDefaultQuality, normalizeOutputProfileId, normalizeQuality } from "./outputQuality";

export const aspectSizeOptions: OutputOption[] = [
  { value: AUTO_OPTION, label: "自动", aspect: AUTO_OPTION, width: 0, height: 0 },
  ...Object.entries(aspectLabels).map(([aspect, label]) => {
    const [width, height] = aspect.split(":").map((value) => Number.parseInt(value, 10));
    return { value: aspect, label, aspect, width, height };
  }),
];

export const pixelAutoOption: OutputOption = { value: AUTO_OPTION, label: "自动", aspect: AUTO_OPTION, quality: AUTO_OPTION, width: 0, height: 0 };
export const pixelPresetOptions = Object.entries(pixelSizeValues).flatMap(([quality, entries]) => (
  entries.map(([aspect, value]) => {
    const [width, height] = parsePixelSize(value) || [0, 0];
    return {
      value,
      label: `${aspectLabels[aspect]} · ${value}`,
      aspect,
      quality,
      width,
      height,
    };
  })
));

export const pixelOptionsByQuality = new Map<string, OutputOption[]>([
  [AUTO_OPTION, [pixelAutoOption]],
  ["standard", [pixelAutoOption, ...pixelPresetOptions.filter((option) => option.quality === "standard")]],
  ["hd", [pixelAutoOption, ...pixelPresetOptions.filter((option) => option.quality === "hd")]],
  ["4k", [pixelAutoOption, ...pixelPresetOptions.filter((option) => option.quality === "4k")]],
]);

export const pixelByValue = new Map([pixelAutoOption, ...pixelPresetOptions].map((option) => [option.value, option]));
export const pixelByAspectAndQuality = new Map<string, Map<string, OutputOption>>();
Object.keys(aspectLabels).forEach((aspect) => {
  const byQuality = new Map<string, OutputOption>();
  ["standard", "hd", "4k"].forEach((quality) => {
    const option = pixelOptionsByQuality.get(quality)?.find((item) => item.aspect === aspect);
    if (option) byQuality.set(quality, option);
  });
  pixelByAspectAndQuality.set(aspect, byQuality);
});

export function getDefaultSizeOption() {
  return AUTO_OPTION;
}

export function inferQualityFromSize(value: unknown, fallback: string | null = null, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  const normalized = String(value || "").trim().toLowerCase();
  if (normalized === AUTO_OPTION) return AUTO_OPTION;
  if (profileId === OUTPUT_PROFILE_ASPECT_V1) return normalizeQuality(fallback, null, profileId);
  const pixelSize = parsePixelSize(normalized);
  if (!pixelSize) return normalizeQuality(fallback, null, profileId);
  const longEdge = Math.max(...pixelSize);
  if (longEdge <= 1600) return "standard";
  if (longEdge <= 2800) return "hd";
  return "4k";
}

export function defaultSizeForQuality(quality: unknown, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  const normalizedQuality = normalizeQuality(quality, null, profileId);
  if (profileId === OUTPUT_PROFILE_ASPECT_V1) return normalizedQuality === AUTO_OPTION ? AUTO_OPTION : "9:16";
  if (normalizedQuality === AUTO_OPTION) return AUTO_OPTION;
  return pixelByAspectAndQuality.get("9:16")?.get(normalizedQuality)?.value || AUTO_OPTION;
}

export function normalizeSizeOption(value: unknown, fallback: string | null = null, quality: unknown = null, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  const normalized = String(value || "").trim().toLowerCase();
  if (normalized === AUTO_OPTION) return AUTO_OPTION;
  if (profileId === OUTPUT_PROFILE_ASPECT_V1) {
    return inferAspectFromSize(normalized) || inferAspectFromSize(fallback) || defaultSizeForQuality(quality, profileId);
  }
  const normalizedQuality = normalizeQuality(quality, null, profileId);
  const lookupQuality = normalizedQuality === AUTO_OPTION ? getResolvedDefaultQuality(profileId) : normalizedQuality;
  if (parsePixelSize(normalized)) return normalized;
  if (pixelByValue.has(normalized)) {
    const option = pixelByValue.get(normalized);
    if (option?.value === AUTO_OPTION) return AUTO_OPTION;
    return pixelByAspectAndQuality.get(option?.aspect || "")?.get(lookupQuality)?.value || defaultSizeForQuality(lookupQuality, profileId);
  }
  if (aspectLabels[normalized]) return pixelByAspectAndQuality.get(normalized)?.get(lookupQuality)?.value || defaultSizeForQuality(lookupQuality, profileId);
  const fallbackNormalized = String(fallback || "").trim().toLowerCase();
  if (fallbackNormalized === AUTO_OPTION) return AUTO_OPTION;
  if (parsePixelSize(fallbackNormalized)) return fallbackNormalized;
  if (aspectLabels[fallbackNormalized]) return pixelByAspectAndQuality.get(fallbackNormalized)?.get(lookupQuality)?.value || defaultSizeForQuality(lookupQuality, profileId);
  return defaultSizeForQuality(lookupQuality, profileId);
}

export function getSizeOptions(quality: unknown, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  if (profileId === OUTPUT_PROFILE_ASPECT_V1) return [...aspectSizeOptions];
  return [...(pixelOptionsByQuality.get(normalizeQuality(quality, null, profileId)) || [pixelAutoOption])];
}

export function getSizeOptionsForValue(quality: unknown, value: unknown, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  const options = getSizeOptions(quality, profileId);
  const normalized = normalizeSizeOption(value, getDefaultSizeOption(), quality, profileId);
  if (options.some((option) => option.value === normalized)) return options;
  const pixelSize = profileId === OUTPUT_PROFILE_PIXEL_V1 ? parsePixelSize(normalized) : null;
  if (!pixelSize) return options;
  const [width, height] = pixelSize;
  return [
    ...options,
    {
      value: normalized,
      label: `自定义像素 · ${normalized}`,
      width,
      height,
      quality: normalizeQuality(quality, null, profileId),
    },
  ];
}

export function formatSize(value: unknown, quality: unknown, outputProfileId = DEFAULT_OUTPUT_PROFILE_ID) {
  const profileId = normalizeOutputProfileId(outputProfileId);
  const normalized = normalizeSizeOption(value, defaultSizeForQuality(quality, profileId), quality, profileId);
  if (profileId === OUTPUT_PROFILE_ASPECT_V1) {
    if (normalized === AUTO_OPTION) return "自动";
    const originalPixel = parsePixelSize(value);
    const aspectLabel = aspectLabels[normalized] || normalized;
    return originalPixel ? `${aspectLabel} · ${originalPixel[0]}x${originalPixel[1]}` : aspectLabel;
  }
  return pixelByValue.get(normalized)?.label || normalized;
}

function inferAspectFromSize(value: unknown) {
  const normalized = String(value || "").trim().toLowerCase();
  if (aspectLabels[normalized]) return normalized;
  const pixelSize = parsePixelSize(normalized);
  if (!pixelSize) return null;
  const [width, height] = pixelSize;
  const ratio = width / height;
  return Object.keys(aspectLabels).reduce((best, aspect) => {
    const [aspectWidth, aspectHeight] = aspect.split(":").map((item) => Number.parseInt(item, 10));
    const delta = Math.abs(aspectWidth / aspectHeight - ratio);
    return !best || delta < best.delta ? { aspect, delta } : best;
  }, null as { aspect: string; delta: number } | null)?.aspect || null;
}

File: SCimage/webapp/output_size.py
from __future__ import annotations

from math import floor

from output_profiles import (
    API_EDGE_MULTIPLE,
    API_MAX_EDGE,
    API_MAX_PIXELS,
    API_MIN_PIXELS,
    ASPECT_LABELS,
    ASPECT_SIZE_OPTIONS,
    ASPECT_SIZE_OPTIONS_WITH_AUTO,
    AUTO_OPTION,
    DEFAULT_OUTPUT_PROFILE_ID,
    DEFAULT_QUALITY,
    DEFAULT_SIZE_OPTION,
    OPENAI_SDK_IMAGE_QUALITY_MAP,
    OUTPUT_PROFILE_ASPECT_V1,
    PIXEL_SIZE_BY_ASPECT_AND_QUALITY,
    PIXEL_SIZE_OPTION_MAP,
    PIXEL_SIZE_OPTIONS_BY_QUALITY,
    PIXEL_TIER_LONG_EDGE_MAX,
    QUALITY_4K,
    QUALITY_HD,
    QUALITY_LOW,
    QUALITY_STANDARD,
    QUALITY_TARGET_LONG_EDGES,
    SizeOption,
    parse_pixel_size,
)
from output_quality import default_quality, normalize_output_profile_id, normalize_quality


def default_size_for_quality(
    quality: object,
    *,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized_quality = normalize_quality(quality, output_profile_id=normalized_profile_id)
    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        return AUTO_OPTION if normalized_quality == AUTO_OPTION else "9:16"
    if normalized_quality == AUTO_OPTION:
        return AUTO_OPTION
    return PIXEL_SIZE_BY_ASPECT_AND_QUALITY["9:16"][normalized_quality].value


def infer_quality_from_size(
    value: object,
    *,
    fallback: str = DEFAULT_QUALITY,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized = str(value or "").strip().lower()
    if normalized == AUTO_OPTION:
        return AUTO_OPTION

    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        return normalize_quality(fallback, output_profile_id=normalized_profile_id)

    pixel_size = parse_pixel_size(value)
    if pixel_size is None:
        if normalized in ASPECT_LABELS:
            return normalize_quality(fallback, output_profile_id=normalized_profile_id)
        return normalize_quality(fallback, output_profile_id=normalized_profile_id)

    long_edge = max(pixel_size)
    if long_edge <= PIXEL_TIER_LONG_EDGE_MAX[QUALITY_STANDARD]:
        return QUALITY_STANDARD
    if long_edge <= PIXEL_TIER_LONG_EDGE_MAX[QUALITY_HD]:
        return QUALITY_HD
    return QUALITY_4K


def available_size_options(
    quality: object,
    *,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> tuple[SizeOption, ...]:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        return ASPECT_SIZE_OPTIONS_WITH_AUTO
    return PIXEL_SIZE_OPTIONS_BY_QUALITY[normalize_quality(quality, output_profile_id=normalized_profile_id)]


def normalize_size_value(
    value: object,
    *,
    fallback: str = DEFAULT_SIZE_OPTION,
    quality: object = DEFAULT_QUALITY,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized = str(value or "").strip().lower()
    if normalized == AUTO_OPTION:
        return AUTO_OPTION

    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        aspect = infer_aspect_from_size(normalized)
        if aspect:
            return aspect
        fallback_aspect = infer_aspect_from_size(fallback)
        if fallback_aspect:
            return fallback_aspect
        return default_size_for_quality(quality, output_profile_id=normalized_profile_id)

    lookup_quality = size_lookup_quality(quality, output_profile_id=normalized_profile_id)
    if parse_pixel_size(normalized):
        return normalized
    if normalized in PIXEL_SIZE_OPTION_MAP:
        option = PIXEL_SIZE_OPTION_MAP[normalized]
        if option.value == AUTO_OPTION:
            return AUTO_OPTION
        return PIXEL_SIZE_BY_ASPECT_AND_QUALITY[option.aspect][lookup_quality].value
    if normalized in ASPECT_LABELS:
        return PIXEL_SIZE_BY_ASPECT_AND_QUALITY[normalized][lookup_quality].value

    fallback_normalized = str(fallback or "").strip().lower()
    if fallback_normalized == AUTO_OPTION:
        return AUTO_OPTION
    if parse_pixel_size(fallback_normalized):
        return fallback_normalized
    if fallback_normalized in ASPECT_LABELS:
        return PIXEL_SIZE_BY_ASPECT_AND_QUALITY[fallback_normalized][lookup_quality].value
    return default_size_for_quality(quality, output_profile_id=normalized_profile_id)


def normalize_size_option(
    value: object,
    *,
    fallback: str = DEFAULT_SIZE_OPTION,
    quality: object = DEFAULT_QUALITY,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    return normalize_size_value(
        value,
        fallback=fallback,
        quality=quality,
        output_profile_id=output_profile_id,
    )


def is_supported_size_value(value: object, *, output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID) -> bool:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized = str(value or "").strip().lower()
    if normalized == AUTO_OPTION:
        return True
    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        return bool(infer_aspect_from_size(normalized))
    return bool(parse_pixel_size(normalized) or normalized in ASPECT_LABELS or normalized in PIXEL_SIZE_OPTION_MAP)


def is_supported_size_option(value: object, *, output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID) -> bool:
    return is_supported_size_value(value, output_profile_id=output_profile_id)


def resolve_api_size_value(
    size: object,
    quality: object,
    *,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized_quality = normalize_quality(quality, output_profile_id=normalized_profile_id)
    normalized_size = normalize_size_value(size, quality=normalized_quality, output_profile_id=normalized_profile_id)
    if normalized_size == AUTO_OPTION:
        return AUTO_OPTION

    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        option = next(item for item in ASPECT_SIZE_OPTIONS if item.value == normalized_size)
        target_quality = normalized_quality if normalized_quality != AUTO_OPTION else QUALITY_LOW
        return resolve_api_dimensions(option, QUALITY_TARGET_LONG_EDGES[target_quality])

    effective_quality = infer_quality_from_size(
        normalized_size,
        fallback=normalized_quality,
        output_profile_id=normalized_profile_id,
    )
    return normalize_size_value(
        normalized_size,
        fallback=default_size_for_quality(effective_quality, output_profile_id=normalized_profile_id),
        quality=effective_quality,
        output_profile_id=normalized_profile_id,
    )


def resolve_openai_sdk_quality(
    quality: object,
    *,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized_quality = normalize_quality(quality, output_profile_id=normalized_profile_id)
    return OPENAI_SDK_IMAGE_QUALITY_MAP.get(normalized_quality, QUALITY_LOW)


def resolve_openai_sdk_size_value(
    size: object,
    quality: object,
    *,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized_quality = normalize_quality(quality, output_profile_id=normalized_profile_id)
    normalized_size = normalize_size_value(size, quality=normalized_quality, output_profile_id=normalized_profile_id)
    if normalized_size == AUTO_OPTION:
        return AUTO_OPTION

    aspect = infer_aspect_from_size(normalized_size)
    if not aspect:
        return AUTO_OPTION
    if aspect == "1:1":
        return "1024x1024"

    width_text, height_text = aspect.split(":", 1)
    return "1536x1024" if int(width_text) > int(height_text) else "1024x1536"


def size_label(
    value: object,
    *,
    quality: object = DEFAULT_QUALITY,
    output_profile_id: str = DEFAULT_OUTPUT_PROFILE_ID,
) -> str:
    normalized_profile_id = normalize_output_profile_id(output_profile_id)
    normalized = normalize_size_value(
        value,
        fallback=default_size_for_quality(quality, output_profile_id=normalized_profile_id),
        quality=quality,
        output_profile_id=normalized_profile_id,
    )

    if normalized_profile_id == OUTPUT_PROFILE_ASPECT_V1:
        if normalized == AUTO_OPTION:
            return "自动"
        original_pixel = parse_pixel_size(value)
        if original_pixel:
            return f"{ASPECT_LABELS.get(normalized, normalized)} · {original_pixel[0]}x{original_pixel[1]}"
        return ASPECT_LABELS.get(normalized, normalized)

    option = PIXEL_SIZE_OPTION_MAP.get(normalized)
    if option:
        return option.label
    return normalized


def infer_aspect_from_size(value: object) -> str | None:
    normalized = str(value or "").strip().lower()
    if normalized in ASPECT_LABELS:
        return normalized
    pixel_size = parse_pixel_size(normalized)
    if pixel_size is None:
        return None
    width, height = pixel_size
    ratio = width / height
    best_aspect = None
    best_delta = None
    for aspect in ASPECT_LABELS:
        aspect_width, aspect_height = map(int, aspect.split(":", 1))
        aspect_ratio = aspect_width / aspect_height
        delta = abs(aspect_ratio - ratio)
        if best_delta is None or delta < best_delta:
            best_delta = delta
            best_aspect = aspect
    return best_aspect


def size_lookup_quality(quality: object, *, output_profile_id: str) -> str:
    normalized_quality = normalize_quality(quality, output_profile_id=output_profile_id)
    return default_quality(output_profile_id) if normalized_quality == AUTO_OPTION else normalized_quality


def resolve_api_dimensions(option: SizeOption, target_long_edge: int) -> str:
    width_ratio = option.width
    height_ratio = option.height
    ratio = width_ratio / height_ratio
    if width_ratio >= height_ratio:
        width = floor_to_multiple(target_long_edge)
        height = floor_to_multiple(width / ratio)
    else:
        height = floor_to_multiple(target_long_edge)
        width = floor_to_multiple(height * ratio)

    while width * height > API_MAX_PIXELS:
        if width >= height:
            width -= API_EDGE_MULTIPLE
            height = floor_to_multiple(width / ratio)
        else:
            height -= API_EDGE_MULTIPLE
            width = floor_to_multiple(height * ratio)

    while width * height < API_MIN_PIXELS:
        if width >= height:
            width += API_EDGE_MULTIPLE
            height = floor_to_multiple(width / ratio)
        else:
            height += API_EDGE_MULTIPLE
            width = floor_to_multiple(height * ratio)
        if width > API_MAX_EDGE or height > API_MAX_EDGE:
            break

    width = min(API_MAX_EDGE, max(API_EDGE_MULTIPLE, width))
    height = min(API_MAX_EDGE, max(API_EDGE_MULTIPLE, height))
    return f"{width}x{height}"


def floor_to_multiple(value: float, multiple: int = API_EDGE_MULTIPLE) -> int:
    return max(multiple, int(floor(value / multiple) * multiple))
